src/CoWSwapEthFlow.sol

```diff
@@ -68,4 +77,47 @@ contract CoWSwapEthFlow is CoWSwapOnchainOrders, ICoWSwapEthFlow {
 
         orders[orderHash] = onchainData;
     }
+
+    /// @inheritdoc ICoWSwapEthFlow
+    function deleteOrder(EthFlowOrder.Data calldata order) external {
+        GPv2Order.Data memory cowSwapOrder = order.toCoWSwapOrder(
+            wrappedNativeToken
+        );
+        bytes32 orderHash = cowSwapOrder.hash(cowSwapDomainSeparator);
+
+        EthFlowOrder.OnchainData memory orderData = orders[orderHash];
+
+        if (
+            orderData.owner == EthFlowOrder.INVALIDATED_OWNER ||
+            orderData.owner == EthFlowOrder.NO_OWNER ||
+            (// solhint-disable-next-line not-rely-on-time
+            orderData.validTo >= block.timestamp &&
+                orderData.owner != msg.sender)
+        ) {
+            revert NotAllowedToDeleteOrder(orderHash);
+        }
+
+        orders[orderHash].owner = EthFlowOrder.INVALIDATED_OWNER;
+
+        bytes memory orderUid = new bytes(GPv2Order.UID_LENGTH);
+        orderUid.packOrderUidParams(
+            orderHash,
+            address(this),
+            cowSwapOrder.validTo
+        );
+        uint256 freedAmount = cowSwapOrder.sellAmount -
+            cowSwapSettlement.filledAmount(orderUid);
+
+        // Using low level calls to perform the transfer avoids setting arbitrary limits to the amount of gas used in a
+        // call. The drawback is allowing reentrancy from this point in the contract. Note that the current order was
+        // already invalidated and cannot be claimed again. TODO: ensure that this is robust under order updates when
+        // implemented.
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success, ) = payable(orderData.owner).call{value: freedAmount}(
```

Comment:
[Here](https://github.com/cowprotocol/ethflowcontract/compare/order-cancellation...non-reentrant-functions?expand=1) I implemented reentrancy guards naively and it seems that it costs about 2500 gas more to make the two involved functions non reentrant:

```
│ Function Name                                                                ┆ min             ┆ avg   ┆ median ┆ max   ┆ # calls │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ createOrder                                                                  ┆ 263             ┆ 30790 ┆ 35795  ┆ 35795 ┆ 16      │
│ createOrder                                                                  ┆ 5316            ┆ 33414 ┆ 38166  ┆ 38166 ┆ 16      │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ deleteOrder                                                                  ┆ 5471            ┆ 27606 ┆ 38726  ┆ 38748 ┆ 11      │
│ deleteOrder                                                                  ┆ 8527            ┆ 29139 ┆ 39103  ┆ 39125 ┆ 11      │
```

I think we should try to avoid these changes as much as possible to save on gas. However, I think it would be reasonable to add the reentrancy guard right now and remove it at a later stage when we are more confident on how the final contract looks like and we can remove it safely. Maybe now I'm slightly in favor of the latter option but I'm ok with both.

Reply:
>remove it at a later stage when we are more confident on how the final contract looks like and we can remove it safely.

let's not put in the work now. We can make an issue about things that we wanna check for audit. We can then still later on decided to put it in, all good.

yeah, I also agree that 2500 is quite heavy if it is not really needed

Reply:
I added an [issue comment](https://github.com/cowprotocol/ethflowcontract/issues/4#issuecomment-1228469060) with the things to check before the audit.